crates/core/src/factories/core_fetch/test.rs

```diff
@@ -0,0 +1,409 @@
+        ..Default::default()
+    };
+    let mock_transport = MockTransport::new(false);
+    let fetch = CoreFetch::new(
+        config.clone(),
+        SpaceId::from(bytes::Bytes::new()),
+        peer_store,
+        mock_transport.clone(),
+    );
+    let agent_id = random_agent_id();
+
+    fetch
+        .0
+        .state
+        .lock()
+        .unwrap()
+        .cool_down_list
+        .add_agent(agent_id.clone());
+
+    assert!(fetch
+        .0
+        .state
+        .lock()
+        .unwrap()
+        .cool_down_list
+        .is_agent_cooling_down(&agent_id));
+
+    // Wait for the cool-down interval + 1 ms to avoid flakiness.
+    tokio::time::sleep(Duration::from_millis(config.cool_down_interval_ms + 1))
+        .await;
+
+    assert!(!fetch
+        .0
+        .state
+        .lock()
+        .unwrap()
+        .cool_down_list
+        .is_agent_cooling_down(&agent_id));
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn multi_op_fetch_from_multiple_unresponsive_agents() {
+    let builder = Arc::new(default_builder());
+    let peer_store = builder.peer_store.create(builder.clone()).await.unwrap();
+    let config = CoreFetchConfig::default();
+    let mock_transport = MockTransport::new(true);
+    let fetch = CoreFetch::new(
+        config.clone(),
+        SpaceId::from(bytes::Bytes::new()),
+        peer_store,
+        mock_transport.clone(),
+    );
+
+    let op_list_1 = create_op_list(10);
+    let agent_1 = random_agent_id();
+    let op_list_2 = create_op_list(20);
+    let agent_2 = random_agent_id();
+    let op_list_3 = create_op_list(30);
+    let agent_3 = random_agent_id();
+
+    // Add all ops to the queue.
+    fetch
+        .add_ops(op_list_1.clone(), agent_1.clone())
+        .await
+        .unwrap();
+    fetch
+        .add_ops(op_list_2.clone(), agent_2.clone())
+        .await
+        .unwrap();
+    fetch
+        .add_ops(op_list_3.clone(), agent_3.clone())
+        .await
+        .unwrap();
+
+    // Wait for one request for each agent.
+    let expected_agents = [agent_1, agent_2, agent_3];
+    tokio::time::timeout(Duration::from_millis(100), async {
+        loop {
+            let requests_sent =
+                mock_transport.requests_sent.lock().await.clone();
+            let request_destinations = requests_sent
+                .iter()
+                .map(|(_, agent_id)| agent_id)
+                .collect::<Vec<_>>();
+            if expected_agents
+                .iter()
+                .all(|agent| request_destinations.contains(&agent))
+            {
+                // Check all agents are on cool_down_list.
+                let cool_down_list =
+                    &mut fetch.0.state.lock().unwrap().cool_down_list;
+                if expected_agents
+                    .iter()
+                    .all(|agent| cool_down_list.is_agent_cooling_down(agent))
+                {
+                    break;
+                }
+            }
+        }
```

Comment:
this is another busy loop

Reply:
lol, I have a deja-vu!
Yes, they were at 10 ms before and that was far too long, so I thought it doesn't make sense to use any sleeps. I'm not used to such low fractions of time, in the Holochain repo it doesn't matter whether you wait for a second or two. =)

Reply:
hehe, yea, you could even use `tokio::task::yield_now().await`... really we just want to give the other task a chance to trigger, because when a task gets cpu time, it'll be optimized to get a mutex lock over and over, but the processing tasks won't get a chance to change anything in the mean time.